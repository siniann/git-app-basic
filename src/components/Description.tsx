import React, { Component } from 'react';
import { TextField, Button } from '@material-ui/core';
import { descriptionRegex } from '../helper/regularExpression';

interface DescriptionState {
  description: string;
  descriptionError: string;
}

interface DescriptionProps {
  saveDescription: Function;
}
class Description extends Component<DescriptionProps, DescriptionState> {
  constructor(props) {
    super(props);
    this.state = {
      description: '',
      descriptionError: '',
    };
  }

  changeDescription = (event) => {
    const isValidSyntax = descriptionRegex(event.target.value);
    if (isValidSyntax) {
      this.setState({ descriptionError: '' });
    } else {
      this.setState({ descriptionError: 'Description should start with either Add,Correct,Change,Remove or Delete' });
    }
    if (event.target.value.length > 100) {
      this.setState({ descriptionError: 'Length should not exceed 100 characters' });
    }
    this.setState({ description: event.target.value });
  };

  saveDescription = () => {
    this.props.saveDescription(this.state.description);
  }

  render() {
    const { descriptionError, description } = this.state;


    return (
      <div>
        <br />
        <i>There are changes in the file. Please add a valid description to change to save it</i>
        <TextField
          fullWidth
          id="change-description"
          value={description}
          label="Description"
          margin="normal"
          variant="outlined"
          placeholder="Add a description..."
          onChange={this.changeDescription}
        />
        {descriptionError ? (
          <div id="description-error">
            {descriptionError}
          </div>
        )
          : null}
        <Button
          disabled={descriptionError !== '' || description === ''}
          fullWidth
          id="btn-saveDescription"
          variant="contained"
          color="default"
          onClick={this.saveDescription}
          size="small"
        >
            save description
        </Button>
      </div>
    );
  }
}

export default Description;
