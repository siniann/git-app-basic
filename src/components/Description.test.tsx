import React from 'react';
import { shallow } from 'enzyme';
import Description from './Description';

describe('Description component', () => {
  const mockFunction = jest.fn();

  let descriptionInstance;
  let descriptionShallow;

  beforeEach(() => {
    jest.clearAllMocks();
    descriptionShallow = shallow(<Description saveDescription={mockFunction} />);
  });

  it('should render properly ', () => {
    expect(descriptionShallow).toMatchSnapshot();
  });

  it('should render properly ', () => {
    descriptionShallow.setState({ descriptionError: 'Some error message' });

    expect(descriptionShallow.state().descriptionError).toEqual('Some error message');
    expect(descriptionShallow.state().description).toEqual('');
    expect(descriptionShallow).toMatchSnapshot();
  });

  it('should not render description error div when no description error is passes ', () => {
    descriptionShallow.setState({
      description: 'Add some valid desc',
      descriptionError: 'Error message: some invalid desc',
    });

    expect(descriptionShallow).toBeDefined();
    expect(descriptionShallow.find('#description-error')).toHaveLength(1);
  });

  it('should render description error div ', () => {
    descriptionShallow.setState({
      description: 'invalid decsc',
      descriptionError: 'some error',
    });
    expect(descriptionShallow.find('#description-error')).toHaveLength(1);
  });
  it('should click the button', () => {
    descriptionShallow.find('#btn-saveDescription').simulate('click');

    expect(mockFunction).toHaveBeenCalled();
  });
  it('should handle changes in description', () => {
    descriptionShallow.find('#change-description').simulate('change', { target: { value: 'Hello Annika' } });
    expect(descriptionShallow.state().descriptionError).toEqual('Description should start with either Add,Correct,Change,Remove or Delete');
  });
  it('should handle changes in description', () => {
    descriptionShallow.find('#change-description').simulate('change', { target: { value: 'Add valid desc' } });
    expect(descriptionShallow.state().descriptionError).toEqual('');
  });

  it('should handle length in description', () => {
    descriptionShallow.find('#change-description').simulate('change', { target: { value: 'Add valid desc valid desc valid desc valid desc valid desc valid desc valid desc valid desc valid desc valid desc valid desc valid desc' } });
    expect(descriptionShallow.state().descriptionError).toEqual('Length should not exceed 100 characters');
  });
});
