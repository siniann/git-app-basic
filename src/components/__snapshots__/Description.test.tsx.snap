// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Description component should render properly  1`] = `ShallowWrapper {}`;

exports[`Description component should render properly  2`] = `ShallowWrapper {}`;
